src/CPUActivityAgent.cpp

```diff
@@ -299,34 +241,55 @@ namespace geopm
-                 std::advance(it, 2)) {
-
-                auto uncore_freq = (*it);
-                auto max_mem_bw = *(it + 1);
-                if (!std::isnan(uncore_freq)) {
-                    // Not valid to have NAN max mem bw for uncore freq.
-                    GEOPM_DEBUG_ASSERT(!std::isnan(max_mem_bw),
-                                       "mapped CPU_UNCORE_FREQUENCY with no max memory bandwidth assigned.");
-                    m_qm_max_rate[uncore_freq] = max_mem_bw;
-                }
-            }
+        // Calculate new frequency range values
+        double f_core_range = m_freq_core_max - m_freq_core_efficient;
+        double f_uncore_range = m_freq_uncore_max - m_freq_uncore_efficient;
 
-            if (m_qm_max_rate.empty()) {
-                throw Exception("CPUActivityAgent::" + std::string(__func__) +
-                                "(): CPUActivityAgent policy did not contain" +
-                                " memory bandwidth characterization.",
-                                GEOPM_ERROR_INVALID, __FILE__, __LINE__);
-            }
+        double phi = in_policy[M_POLICY_CPU_PHI];
+
+        // Default phi = 0.5 case is full fe to fmax range
+        // Core
+        m_resolved_f_core_max = m_freq_core_max;
+        m_resolved_f_core_efficient = m_freq_core_efficient;
+        // Uncore
+        m_resolved_f_uncore_max = m_freq_uncore_max;
+        m_resolved_f_uncore_efficient = m_freq_uncore_efficient;
+
+        // If phi is not 0.5 we move into the energy or performance biased behavior
+        if (phi > 0.5) {
+            // Energy Biased.  Scale F_max down to F_efficient based upon phi value
+            // Active region phi usage
+            m_resolved_f_core_max = std::max(m_freq_core_efficient, m_freq_core_max -
+                                                                    f_core_range *
+                                                                    (phi-0.5) / 0.5);
+
+            m_resolved_f_uncore_max = std::max(m_freq_uncore_efficient, m_freq_uncore_max -
+                                                                        f_uncore_range *
+                                                                        (phi-0.5) / 0.5);
         }
+        else if (phi < 0.5) {
+            // Perf Biased.  Scale F_efficient up to F_max based upon phi value
+            // Active region phi usage
+            m_resolved_f_core_efficient = std::min(m_freq_core_max, m_freq_core_efficient +
+                                                                    f_core_range *
+                                                                    (0.5-phi) / 0.5);
+
+            m_resolved_f_uncore_efficient = std::min(m_freq_uncore_max, m_freq_uncore_efficient +
+                                                                        f_uncore_range *
+                                                                        (0.5-phi) / 0.5);
```

Comment:
Same idea here, save the `(0.5-phi) / 0.5)` into it's own variable, and substitute it for both of these expressions.